cli/flox/src/commands/init/python.rs

```diff
@@ -0,0 +1,664 @@
+    Invalid(Error),
+    /// Provider not found
+    NotFound,
+}
+
+impl<P: Provider + 'static> From<Result<Option<P>>> for Provide<Box<dyn Provider>> {
+    fn from(result: Result<Option<P>>) -> Self {
+        match result {
+            Ok(Some(provider)) => Provide::Found(Box::new(provider)),
+            Ok(None) => Provide::NotFound,
+            Err(err) => Provide::Invalid(err),
+        }
+    }
+}
+
+trait Provider: Debug {
+    fn describe_provider(&self) -> Cow<'static, str>;
+
+    fn describe_reason(&self) -> Cow<'static, str>;
+
+    fn describe_customization(&self) -> Cow<'static, str>;
+
+    fn compatible(&self) -> bool {
+        true
+    }
+
+    fn get_init_customization(&self) -> InitCustomization;
+}
+
+/// Information gathered from a pyproject.toml file for poetry
+/// <https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#configuring-setup-py>
+#[derive(Debug)]
+struct PoetryPyProject {
+    /// Provided python version
+    ///
+    /// [ProvidedVersion::Compatible] if a version compatible with the requirement
+    /// `tools.poetry.dependencies.python` in the pyproject.toml was found in the catalogs.
+    ///
+    ///  <https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups>
+    ///
+    /// [ProvidedVersion::Default] if no compatible version was found, but a default version was found.
+    provided_python_version: ProvidedVersion,
+
+    /// Version of poetry found in the catalog
+    poetry_version: String,
+}
+
+impl PoetryPyProject {
+    fn detect(path: &Path, flox: &Flox) -> Result<Option<Self>> {
+        debug!("Detecting poetry pyproject.toml at {:?}", path);
+
+        let pyproject_toml = path.join("pyproject.toml");
+
+        if !pyproject_toml.exists() {
+            debug!("No pyproject.toml found at {:?}", path);
+            return Ok(None);
+        }
+
+        let content = std::fs::read_to_string(&pyproject_toml)?;
+
+        let toml = toml_edit::Document::from_str(&content)?;
+
+        // poetry _requires_ `tool.poetry.dependencies.python` to be set [1],
+        // so we do not resolve a default version here if the key is missing.
+        //[1]: <https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups>
+        let Some(poetry) = toml.get("tool").and_then(|tool| tool.get("poetry")) else {
+            return Ok(None);
+        };
+
+        let required_python_version = poetry
+            .get("dependencies")
+            .and_then(|dependencies| dependencies.get("python"))
+            .map(|python| python.as_str().context("expected a string"))
+            .transpose()?
+            .ok_or_else(|| {
+                anyhow!("No python version specified at 'tool.poetry.dependencies.python'")
+            })?
+            .to_string();
+
+        let provided_python_version = 'version: {
+            let compatible =
+                get_default_package("python3", Some(required_python_version.clone()), flox)?;
```

Comment:
I think this is similar to the question regarding how we resolve versions that i alluded to in the npm hook PR.

Basically, do we want to use nix' python versioning OR depend on versioning through the history of the `python3` package.
This is currently trying to satisfy the version constraint by choosing any python3 package.
maybe we should constrain this more to just search matching `pname`s, or decide to just search _the_ `python3` package and bet on the catalog to provide versions for it.

Reply:
I don't think I'm quite understanding. `Query::new` uses `match_name_or_rel_path`, so we shouldn't ever pass that a version. We should only use that for "What package do users get if they type python3?" But python3@user-provided-input is a bad idea because it will return things that aren't python at all (at least until we do https://github.com/flox/flox/issues/816)

We can consider using either pname or rel_path, but partial match on rel_path is wrong